File: Cpp/Logger/Logger.cpp
//---------------------------------------------------//
//                    MIT License                    //
// Copyright @ 2018-2019 Tony Su All Rights Reserved //
//        https://github.com/peitaosu/Logger         //
//---------------------------------------------------//

#include "Logger.h"

Logger::Logger()
{
}

Logger::~Logger()
{
}

Logger* Logger::Instance() {
    if (!_logger) _logger = new Logger();
    return _logger;
}

void Logger::LoadConfig() {
    LoadConfig(this->_defaultConfig);
}

void Logger::LoadConfig(std::string config) {
    this->_config = LogConfig(config);
    if (this->_config.GetLogRoot().MinLevel.Value == "VERBO") this->_minimalLevel = LogEventLevel::VERBO;
    if (this->_config.GetLogRoot().MinLevel.Value == "DEBUG") this->_minimalLevel = LogEventLevel::DEBUG;
    if (this->_config.GetLogRoot().MinLevel.Value == "INFOR") this->_minimalLevel = LogEventLevel::INFOR;
    if (this->_config.GetLogRoot().MinLevel.Value == "WARNN") this->_minimalLevel = LogEventLevel::WARNN;
    if (this->_config.GetLogRoot().MinLevel.Value == "ERROR") this->_minimalLevel = LogEventLevel::ERROR;
    if (this->_config.GetLogRoot().MinLevel.Value == "FATAL") this->_minimalLevel = LogEventLevel::FATAL;
    for (auto& appender : this->_config.GetLogAppenders()) {
        if (appender.Type == "FileAppender" && appender.Enabled) {
            this->_logPath = appender.File.Path;
            if (appender.File.AppendTo == "false") {
                this->_appendTo = false;
            }
        }
    }
}

void Logger::EnableAppender(std::string appenderType) {
    EnableAppender(appenderType, true);
}

void Logger::EnableAppender(std::string appenderType, bool enable) {
    std::list<LogAppender> appenders = this->_config.GetLogAppenders();
    for (auto& appender : appenders) {
        if (appender.Type.compare(appenderType) == 0) {
            appender.Enabled = enable;
        }
    }
    this->_config.UpdateLogAppenders(appenders);
}

void Logger::SetLogPath(std::string log) {
    this->_logPath = log;
}

void Logger::SetLogAppendTo(bool appendTo)
{
    this->_appendTo = appendTo;
}

bool Logger::IsEnabled(LogEventLevel level) {
    if(level < this->_minimalLevel) return false;
    return true;
}

void Logger::Write(LogEvent logEvent) {
    for (auto& appender : this->_config.GetLogAppenders()) {
        if (!appender.Enabled) continue;
        std::string log;
        std::string level;
        switch (logEvent.GetLevel())
        {
        case LogEventLevel::VERBO:
            level = "VERBO";
            break;
        case LogEventLevel::DEBUG:
            level = "DEBUG";
            break;
        case LogEventLevel::INFOR:
            level = "INFOR";
            break;
        case LogEventLevel::WARNN:
            level = "WARNN";
            break;
        case LogEventLevel::ERROR:
            level = "ERROR";
            break;
        case LogEventLevel::FATAL:
            level = "FATAL";
            break;
        default:
            level = "VERBO";
            break;
        }
        std::string fmt = appender.Pattern.Value;
        std::string time_fmt, level_fmt, log_fmt, whole_fmt;
        std::regex pieces_regex(".*\\{(.*?)\\}.*\\{(.*?)\\}.*\\{(.*?)\\}.*");
        std::smatch pieces_match;
        if (std::regex_match(fmt, pieces_match, pieces_regex)) {
            time_fmt = pieces_match[1].str();
            if (time_fmt == "0") time_fmt = "%Y/%m/%d %H:%M:%S";
            level_fmt = pieces_match[2].str();
            if (level_fmt == "1") level_fmt = "%s";
            log_fmt = pieces_match[3].str();
            if (log_fmt == "2") log_fmt = "%s";
            std::ostringstream time_string;
            time_string << std::put_time(&logEvent.GetTimestamp(), time_fmt.c_str());
            log = fmt;
            size_t index = 0;
            index = log.find("{" + pieces_match[1].str() + "}", index);
            log.replace(index, pieces_match[1].str().size() + 2, time_string.str());
            index = log.find("{" + pieces_match[2].str() + "}", index);
            log.replace(index, pieces_match[2].str().size() + 2, level);
            index = log.find("{" + pieces_match[3].str() + "}", index);
            log.replace(index, pieces_match[3].str().size() + 2, logEvent.GetMessage());
        }
        else
        {
            std::ostringstream time_string;
            time_string << std::put_time(&logEvent.GetTimestamp(), "%Y/%m/%d %H:%M:%S");
            log = time_string.str() + ": [" + level + "] " + logEvent.GetMessage();
        }
        if (logEvent.GetException().GetSummary() != "") {
            log = log + "\n";
            log += logEvent.GetException().GetSummary();
        }
        if (appender.Type == "FileAppender") {
            int mode = std::ios::out;
            if (this->_appendTo)
            {
                    mode += std::ios::app;
            }
            std::ofstream file(this->_logPath, mode);
            file << log + "\n";
            file.close();
        }
        else if (appender.Type == "ConsoleAppender") {
            std::cout << log << std::endl;
        }
        else if (appender.Type == "ColoredConsoleAppender") {
            bool color_not_found = true;
            for (auto& color : appender.Colors) {
                if (color.Level == level && color.ForeColor != "") {
                    std::cout << this->_logColor.GetColorByName(color.ForeColor) << log << this->_logColor.GetColorByName("ENDC") << std::endl;
                    color_not_found = false;
                }
            }
            if(color_not_found) std::cout << log << std::endl;
        }
    }
}

void Logger::Write(LogEventLevel level, std::string message) {
    LogEventException exception;
    if (IsEnabled(level))
        Write(LogEvent(level, message, exception));
}

void Logger::Write(LogEventLevel level, std::string message, LogEventException exception) {
    if (IsEnabled(level))
        Write(LogEvent(level, message, exception));
}

void Logger::Verbose(std::string message) {
    Write(LogEventLevel::VERBO, message);
}

void Logger::Verbose(std::string message, LogEventException exception) {
    Write(LogEventLevel::VERBO, message, exception);
}

void Logger::Debug(std::string message) {
    Write(LogEventLevel::DEBUG, message);
}

void Logger::Debug(std::string message, LogEventException exception) {
    Write(LogEventLevel::DEBUG, message, exception);
}

void Logger::Information(std::string message) {
    Write(LogEventLevel::INFOR, message);
}

void Logger::Information(std::string message, LogEventException exception) {
    Write(LogEventLevel::INFOR, message, exception);
}

void Logger::Warning(std::string message) {
    Write(LogEventLevel::WARNN, message);
}

void Logger::Warning(std::string message, LogEventException exception) {
    Write(LogEventLevel::WARNN, message, exception);
}

void Logger::Error(std::string message) {
    Write(LogEventLevel::ERROR, message);
}

void Logger::Error(std::string message, LogEventException exception) {
    Write(LogEventLevel::ERROR, message, exception);
}

void Logger::Fatal(std::string message) {
    Write(LogEventLevel::FATAL, message);
}

void Logger::Fatal(std::string message, LogEventException exception) {
    Write(LogEventLevel::FATAL, message, exception);
}

Logger* Logger::_logger = nullptr;

File: CSharp/Logger/Logger.cs
﻿using System;
using System.Collections.Generic;
using System.IO;
using System.Xml;
using System.Xml.Schema;

namespace Logging
{
    public class Logger
    {
        public LogAppenderConfig FileLog = new LogAppenderConfig();
        public LogAppenderConfig ConsoleLog = new LogAppenderConfig();
        public LogAppenderConfig ColoredConsoleLog = new LogAppenderConfig();
        public LogRootConfig RootConfig = new LogRootConfig();

        private static Logger _logger = null;
        private static LogConfig _logConfig = null;
        private static readonly object _sync = new object();
        private StreamWriter _fileWriter;
        private string _defaultConfig = "LogConfig.xml";
        private string _config = null;

        public Logger(){}

        public void LoadConfig(string config=null)
        {
            _logConfig = new LogConfig();
            if(config == null)
            {
                try
                {
                    _logConfig.Load(Path.Combine(Directory.GetCurrentDirectory(), _defaultConfig));
                }
                catch (FileNotFoundException)
                {
                    _logConfig.Load(Path.Combine(Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location), _defaultConfig));
                }
            }
            else
            {
                this._config = config;
                _logConfig.Load(config);
            }
            List<string> refedAppenders = new List<string>();
            foreach (LogConfig.RootAppenderRef appenderRef in _logConfig.Root.RootAppenderRefs)
            {
                refedAppenders.Add(appenderRef.Ref);
            }
            LogEventLevel minimumLevel = (LogEventLevel)Enum.Parse(typeof(LogEventLevel), _logConfig.Root.MinLevel.Value, true);
            RootConfig = new LogRootConfig { MinimumLevel = minimumLevel, RefedAppenders = refedAppenders };

            foreach (LogConfig.LogAppender appender in _logConfig.Appenders)
            {
                if (!RootConfig.RefedAppenders.Contains(appender.Name))
                {
                    continue;
                }
                switch (appender.Type)
                {
                    case "FileAppender":
                        FileLog = new LogAppenderConfig { IsEnabled = true, LogPath = appender.File.Path, LogAppendTo = (appender.File.AppendTo.Equals("true", StringComparison.OrdinalIgnoreCase)), LogPattern = appender.Pattern.Value, ColoredConsoleMapping = null };
                        if (!FileLog.LogAppendTo)
                        {
                            using (_fileWriter = new StreamWriter(FileLog.LogPath, false)) { }
                        }
                        break;
                    case "ConsoleAppender":
                        ConsoleLog = new LogAppenderConfig { IsEnabled = true, LogPath = null, LogAppendTo = false, LogPattern = appender.Pattern.Value, ColoredConsoleMapping = null };
                        break;
                    case "ColoredConsoleAppender":
                        Dictionary<LogEventLevel, Dictionary<string, string>> levelColorMapping = new Dictionary<LogEventLevel, Dictionary<string, string>>();
                        foreach (LogConfig.AppenderColor color in appender.Colors)
                        {
                            Dictionary<string, string> colorMapping = new Dictionary<string, string>();
                            if (color.ForeColor != null) colorMapping.Add("ForeColor", color.ForeColor);
                            if (color.BackColor != null) colorMapping.Add("BackColor", color.BackColor);
                            levelColorMapping.Add((LogEventLevel)Enum.Parse(typeof(LogEventLevel), color.Level, true), colorMapping);
                        }
                        ColoredConsoleLog = new LogAppenderConfig { IsEnabled = true, LogPath = null, LogAppendTo = false, LogPattern = appender.Pattern.Value, ColoredConsoleMapping = levelColorMapping };
                        break;
                    default:
                        break;
                }
            }
        }

        public static Logger Instance
        {
            get
            {
                if (null == _logger)
                {
                    lock (_sync)
                    {
                        if (null == _logger)
                        {
                            _logger = new Logger();
                            _logger.LoadConfig();
                        }
                    }
                }
                return _logger;
            }
        }

        public bool ValidateSchema(string xsdPath)
        {
            XmlDocument xml = new XmlDocument();
            xml.Load(this._config);
            xml.Schemas.Add(null, xsdPath);

            try
            {
                xml.Validate(null);
            }
            catch (XmlSchemaValidationException)
            {
                throw;
            }
            return true;
        }

        public void EnableAppender(string appenderType, bool enable = true)
        {
            switch (appenderType)
            {
                case "FileAppender":
                    FileLog.IsEnabled = enable;
                    break;
                case "ConsoleAppender":
                    ConsoleLog.IsEnabled = enable;
                    break;
                case "ColoredConsoleAppender":
                    ColoredConsoleLog.IsEnabled = enable;
                    break;
                default:
                    break;
            }
        }

        public void SetLogPath(string log)
        {
            this.FileLog.LogPath = log;
        }

        public void SetLogAppendTo(bool appendTo)
        {
            this.FileLog.LogAppendTo = appendTo;
        }

        public bool IsEnabled(LogEventLevel level)
        {
            if ((int)level < (int)RootConfig.MinimumLevel)
                return false;
            return true;
        }

        public void Write(LogEvent logEvent)
        {
            if (logEvent == null) return;
            if (!IsEnabled(logEvent.Level)) return;
            if (FileLog.IsEnabled)
            {
                using (_fileWriter = new StreamWriter(FileLog.LogPath, true))
                {
                    _fileWriter.WriteLine(FileLog.LogPattern, logEvent.Timestamp, logEvent.Level, logEvent.Message);
                    if (logEvent.Exception != null)
                    {
                        _fileWriter.WriteLine(FileLog.LogPattern, logEvent.Timestamp, logEvent.Level, logEvent.Exception);
                    }
                }
            }
            if (ConsoleLog.IsEnabled)
            {
                Console.WriteLine(ConsoleLog.LogPattern, logEvent.Timestamp, logEvent.Level, logEvent.Message);
                if (logEvent.Exception != null)
                {
                    Console.WriteLine(ConsoleLog.LogPattern, logEvent.Timestamp, logEvent.Level, logEvent.Exception);
                }
            }
            if (ColoredConsoleLog.IsEnabled)
            {
                if (ColoredConsoleLog.ColoredConsoleMapping.ContainsKey(logEvent.Level))
                {
                    if (ColoredConsoleLog.ColoredConsoleMapping[logEvent.Level].ContainsKey("ForeColor"))
                    {
                        Console.ForegroundColor = (ConsoleColor)Enum.Parse(typeof(ConsoleColor), ColoredConsoleLog.ColoredConsoleMapping[logEvent.Level]["ForeColor"]);
                    }
                    if (ColoredConsoleLog.ColoredConsoleMapping[logEvent.Level].ContainsKey("BackColor"))
                    {
                        Console.BackgroundColor = (ConsoleColor)Enum.Parse(typeof(ConsoleColor), ColoredConsoleLog.ColoredConsoleMapping[logEvent.Level]["BackColor"]);
                    }
                }
                Console.WriteLine(ColoredConsoleLog.LogPattern, logEvent.Timestamp, logEvent.Level, logEvent.Message);
                if (logEvent.Exception != null)
                {
                    Console.WriteLine(ColoredConsoleLog.LogPattern, logEvent.Timestamp, logEvent.Level, logEvent.Exception);
                }
                Console.ResetColor();
            }
        }

        public void Write(LogEventLevel level, string message, Exception exception = null)
        {
            if (IsEnabled(level))
                Write(new LogEvent(level, message, exception));
        }

        public void Verbose(string message, Exception exception = null)
        {
            Write(LogEventLevel.VERBO, message, exception);
        }

        public void Debug(string message, Exception exception = null)
        {
#if DEBUG
            Write(LogEventLevel.DEBUG, message, exception);
#endif
        }

        public void Information(string message, Exception exception = null)
        {
            Write(LogEventLevel.INFOR, message, exception);
        }

        public void Warning(string message, Exception exception = null)
        {
            Write(LogEventLevel.WARNN, message, exception);
        }

        public void Error(string message, Exception exception = null)
        {
            Write(LogEventLevel.ERROR, message, exception);
        }

        public void Fatal(string message, Exception exception = null)
        {
            Write(LogEventLevel.FATAL, message, exception);
        }
    }

    public class LogAppenderConfig
    {
        public bool IsEnabled = false;
        public string LogPath = null;
        public bool LogAppendTo = true;
        public string LogPattern = null;
        public Dictionary<LogEventLevel, Dictionary<string, string>> ColoredConsoleMapping = null;
    }

    public class LogRootConfig
    {
        public LogEventLevel MinimumLevel = LogEventLevel.VERBO;
        public List<string> RefedAppenders = null;
    }
}
